### ISSUE TITLE
Landing page for filtering of specific events

### ISSUE THREAD

**Issue by coderesting** [jeff-saupe/ScheduleCleaner#5]
There are currently both english courses in your schedule.
![image](https://user-images.githubusercontent.com/17936518/117129111-93bce400-ad9e-11eb-881e-389f55008edc.png)
It would be cool if only your english course is displayed.

**Comment by jeff-saupe** [comment]
At the moment, this can be done [programmatically](https://github.com/jeff-saupe/NAK-ScheduleCleaner#apply-fixes-for-specific-events).

But for the console dialog, it would be great to show duplicates to the user. If there are multiple modules at the same time (and with the same name), the user should be able to pick the one he/she wants to keep. This must happen before the ICS building part, so the system can filter out the unwanted module.

**Comment by jeff-saupe** [comment]
For the **URL version**, the creation of filters via **parameters** just came up to my mind. 

The following one would **exclude** any event from the ICS file which **contains** the phrase **Dozent XY**.

`https://........./<centuria>-<semester>.ics?type=exclusion&method=contains&text=Dozent XY`

The next one would apply for **updating a title**:

`https://........./<centuria>-<semester>.ics?type=update&method=contains&text=Technische Grundlagen&replaceWith=TGdI`

Any help on this issue is appreciated!! :)

**Comment by jeff-saupe** [comment]
Using a JSON string is definitely a considerable approach and maybe would make the implementation easier. But I think it'd be a difficult solution from a **user perspective**. Even for a non-techie, concatenating parameters to a URL is easier compared to a JSON (considering the correct syntax and so on).

Multiple actions could be handled by simply adding more actions to the URL:
https://........./<centuria>-<semester>.ics?type=exclusion1&method=contains&text=abc%type=exclusion2&method=contains&text=cde

Yes, I know, this would violate the parameter's uniqueness, but in this case it's acceptable I think.

**Comment by coderesting** [comment]
I thought there would still be a script to access the REST API. If you don't plan to include such a script I would agree to make the URL as simple as possible. Anyways, I still think that the order of parameters mustn't be important and would propose
https://........./-.ics?exclude=contains;abc&exclude=contains;def&update=Technische Grundlagen;TGDI
This pattern would make easier for the user to understand which options (contains,text etc) belong to which action (exclude,update).

**Comment by jeff-saupe** [comment]
It's possible to use Spring Boot as framework for REST calls, but tbh it'd be a bit overkill for this purpose. The simplified URL scheme is a good idea tho. 👍 

So an **EventExclusion** with **FixMethod.CONTAINS** would be like this:
- exclude=contains;text

And a **TitleUpdate** with **FixMethod.CONTAINS** would be like this (note the third parameter):
- update=contains;oldTitle;newTitle

As a full URL:
- https://........./-.ics?exclude=contains;text&update=contains;oldTitle;newTitle

**Comment by coderesting** [comment]
Yes, that was my plan.
I agree that spring boot is overkill.
Regex parsing or a smaller library will probably be fine.

**Comment by coderesting** [comment]
To address my original issue I would like to have a fix method removeDuplicates where events with 100% overlap will be removed. This could be done, like you stated earlier, for **specific events** with other methods, but it would be nice to have a generalized solution for this problem. As this change would probably require a lot of rewriting in the fix handling I leave it up to you if such a feature is worth the time. What do you think?

**Comment by jeff-saupe** [comment]
You're probably talking about the duplicate regarding the two English courses. In fact, they seem to be duplicates as they're running at the same time and have the same title. 

But, and here is what I'm worried about, there are additional information for each event:

![image](https://user-images.githubusercontent.com/46161762/118192622-b5753580-b446-11eb-8b21-33f5aba8bd89.png)
![image](https://user-images.githubusercontent.com/46161762/118192635-ba39e980-b446-11eb-9f42-4a0fc5b5a55d.png)

As you can see, there are several fields where the courses can **differ**, such as Dozent, Raum and Anmerkung. If we solely base the removal of duplicates on the title, here 'Englisch', we can't be sure thate we delete the right course for the student.

**Comment by coderesting** [comment]
Well, I haven't thought about the extra information.
Regarding your problem:
Maybe it would be possible to interpret the text of exclude as a regex or provide an additional action to filter with regex.
This approach obviously contain some security risks and  the user probably only needs a wildcard like .* to exclude English.*O'Brien.

**Comment by jeff-saupe** [comment]
I'd like to avoid offering RegEx to the user, as this could also possibly lead to unforeseen problems. But this made me come up with another idea: We could allow multiple Strings as a parameter for an exclusion e.g:

- &exclude=contains;O'Brien;Englisch;String3;String4;....

So EventExclusion would allow an infinite amount of parameters and TitleUpate would allow only two parameters: oldTitle and newTitle.

**Comment by coderesting** [comment]
That approach sounds promising 👍

**Comment by jeff-saupe** [comment]
Alright! If we go with #16, it'd be of course like this:

- &exclude=O'Brien;Englisch;String3;String4;....

**Comment by jeff-saupe** [comment]
With the latest changes on branch **dev**, a major refactoring was done.
- The class `DynamicIcsServer` had been moved to package `de.saupe.jeff.schedulecleaner.environment.impl`.
-  It's now possible to add fixes by calling the method `addFix(Fix fix)` of `ScheduleCleaner`.

**Comment by jeff-saupe** [comment]
I couldn't resist implementing it, sorry.. 😂 But it still needs some testing as parsing the URL has more edge cases than expected. Nevertheless, I hope I've covered all possible cases.

I'm going to merge it into production now for testing and afterwards I'll update the usage section of the README.

**Comment by jeff-saupe** [comment]
I don't know why but Heroku seems to ignore the parameters @coderesting . I've created a branch called **test-environment** where you can see that the URL and paramters are working. Take a look into the classes **Main** and **TestEnvironment**.

Maybe you can investigate this, whenever you have time. No hurry. :)

**Comment by jeff-saupe** [comment]
The issue has been fixed. :) I'm going to write the usage section tomorrow or so and increment the version. **Any testing is welcomed!**

The URL needs to be **encoded**, otherwise it can't be added to Google Calendars, e.g.:
`https://schedule-cleaner.herokuapp.com/cleaned-schedule/a19a_4.ics?exclude=O%27Brien;Englisch&replace=Tech.Grundlagen%20der%20Informatik%202;TGdI`

**Comment by coderesting** [comment]
I think typing the encoded URL completely ruins our user oriented url structure. If we can't get around this issue, I would suggest to add a landing page, where the user can add his fixes and then gets the url. Maybe we can then also revert to the json structure (I'm still not happy with parameter duplicates but we can also keep the current structure if you already implemented it).

**Comment by jeff-saupe** [comment]
Yep, I'm 100% with you - it's not user friendly. For the moment it will work though, as there are multiple URL Encoding tools online. But I think this is a problem we can't get around if using URL.

That's why having a landing page, or any sort of 'building' page would be the best option. And using JSON would definitely be more handy than. Maybe it's possible to keep both structures, depending on the user's preference.
